### PR TITLE
 Add validation exception for old teacher accounts. 

### PR DESCRIPTION
**Pull request by daynew**
A few years ago, we added an email requirement for teacher accounts.
At the time, we did not backfill existing teacher accounts, so we still
have active teacher accounts which don't have emails recorded for them.
These users currently experience pain because the code currently refuses
to update change to their account because the account is currently
invalid. This change makes it so these older accounts will be considered
valid for the time being. There is follow up work to get an email
recorded for the teacher accounts.

### Future work
https://codedotorg.atlassian.net/browse/FND-1130

## Links
- [jira](https://codedotorg.atlassian.net/browse/FND-1018)

## Testing story
- unit tests

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -808,7 +809,31 @@ def password_required?
...
+      if created_at
+        # existing accounts created after the email requirement must have an
+        # email.
+        return  created_at.to_datetime > DATE_TEACHER_EMAIL_REQUIREMENT_ADDED
+      end
+      # new teacher accounts should always require an email
+      return true
+    end
+    # non-teachers are not relevant to this method.
+    return false
```
Comment (Hamms) [inline dashboard/app/models/user.rb]: This is a bit hard to follow. I'd recommend a pattern like this:

```ruby
def teacher_email_required?
  # non-teachers are not relevant to this method.
  return false unless teacher? && purged_at.nil?

  # new teacher accounts should always require an email
  return true unless created_at.present?

  # existing accounts created after the email requirement must have an email.
  # FND-1130: The created_at exception will no longer be required
  # Remove the created_at > '2016-06-14 00:00:00' once all teachers have
  # emails.
  return created_at.to_datetime > DATE_TEACHER_EMAIL_REQUIREMENT_ADDED
end
```

File: dashboard/test/models/user_test.rb [hunk 2]
```diff
@@ -679,10 +679,25 @@ def assert_fails_email_uniqueness_validation(user)
     end
   end
 
-  test "cannot create teacher without email" do
-    assert_does_not_create(User) do
```
Comment (Hamms) [inline dashboard/test/models/user_test.rb:683]: is there a specific reason to move away from the `assert_does_not_create` test?

Comment (daynew) [reply dashboard/test/models/user_test.rb:683]: yes, but that's because I didn't realize `assert_does_not_create` returns the user object. I have brought this assertion back.

**Review by Hamms — APPROVED** [review]
looks great!
